### PR TITLE
[SPARK-26176][SQL] Verify column names for CTAS with `STORED AS`

### PR DESCRIPTION
**Pull request by sujith71955**
## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE**
```scala
scala> sql("set spark.sql.hive.convertMetastoreParquet=false")
scala> sql("CREATE TABLE a STORED AS PARQUET AS SELECT 1 AS `COUNT(ID)`")
Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
```

**AFTER**
```scala
scala> sql("CREATE TABLE a STORED AS PARQUET AS SELECT 1 AS `COUNT(ID)`")
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
```

## How was this patch tested? 
Pass the Jenkins with the newly added test case.

### CODE REVIEW

**Comment by sujith71955** [comment]
retest this please

**Comment by VaibhavDesai** [comment]
**[Test build Jenkins build has started](https://ci-dev.ngcc.nvidia.com/job/vaibhavd_spark/106/testReport)** for PR 24075 at commit [`d73cfb5`](https://github.com/apache/spark/commit/d73cfb51941f99516b7878acace26db35ea72076).

**Comment by VaibhavDesai** [comment]
**[Test build Jenkins build has started](https://ci-dev.ngcc.nvidia.com/job/vaibhavd_spark/106/testReport)** for PR 24075 at commit [`d73cfb5`](https://github.com/apache/spark/commit/d73cfb51941f99516b7878acace26db35ea72076).

**Comment by sujith71955** [comment]
> **[Test build #103390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103390/testReport)** for PR 24075 at commit [`e38e282`](https://github.com/apache/spark/commit/e38e2829ccc98ef1b9924ab452df2e2494d3c355).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

seems to be random failure

**Comment by sujith71955** [comment]
retest this please

**Comment by sujith71955** [comment]
Above failure is not related to this PR. Please review and let me know for any suugestion. Thanks
cc @HyukjinKwon @dongjoon-hyun @cloud-fan

**Comment by sujith71955** [comment]
cc @gatorsmile

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 1]
```diff
@@ -155,7 +155,7 @@ object HiveAnalysis extends Rule[LogicalPlan] {
       CreateTableCommand(tableDesc, ignoreIfExists = mode == SaveMode.Ignore)
 
     case CreateTable(tableDesc, mode, Some(query)) if DDLUtils.isHiveTable(tableDesc) =>
-      DDLUtils.checkDataColNames(tableDesc)
+      DDLUtils.checkDataColNames(tableDesc.copy(schema = query.schema))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: How is it done for data source tables? by another rule?

Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Yes in datasource table scenario, the flow will go through DataSourceAnalysis rule.

Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: And moreover one more problem what i observed is in serde class name defined "parquet.hive.serde.ParquetHiveSerDe" in checkDataColNames() API,  the serde name shall be "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe" for parquet, so i added this serde also in above code as part of checkDataColNames()  API.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @sujith71955 . I updated the PR description slightly. I also take a look at this tonight.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by sujith71955** [comment]
> Thank you for pinging me, @sujith71955 .
> 
> * I updated the PR description slightly and triggered a new testing since there was no successful run until now.
> * In addition, I update this JIRA as an `Improvement` since the previous and new behavior are just the same except raising the better exceptions for UX.

Sure. Thanks :)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: can we unify this check for both data source table and hive serde table?

Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: sure, let me check. thanks for your input.

Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Both are called from different rules, i will check how to unify

Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Handled as part of PreprocessTableCreation rule for CTAS query, please review and let me know for any suggestions. Thanks

**Comment by sujith71955** [comment]
retest this please

**Comment by sujith71955** [comment]
> **[Test build #103568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103568/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

Seems to be failures not related to the PR

**Comment by sujith71955** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by sujith71955** [comment]
Gentle ping @dongjoon-hyun @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -206,6 +206,8 @@ case class PreprocessTableCreation(sparkSession: SparkSession) extends Rule[Logi
         val analyzedQuery = query.get
         val normalizedTable = normalizeCatalogTable(analyzedQuery.schema, tableDesc)
 
+        DDLUtils.checkDataColNames(tableDesc.copy(schema = analyzedQuery.schema))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: did we call this in the else branch?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 3]
```diff
@@ -210,7 +209,7 @@ case class RelationConversions(
       case CreateTable(tableDesc, mode, Some(query))
           if DDLUtils.isHiveTable(tableDesc) && tableDesc.partitionColumnNames.isEmpty &&
             isConvertible(tableDesc) && SQLConf.get.getConf(HiveUtils.CONVERT_METASTORE_CTAS) =>
-        DDLUtils.checkDataColNames(tableDesc)
+        DDLUtils.checkDataColNames(tableDesc.copy(schema = query.schema))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Do we need to call it here?

**Comment by sujith71955** [comment]
Nope, i unified this for ctas queries, if non ctas query still flow is
same, means the validation will be done from HiveAnalysis and
DatasourceAnalysis rule. Shall we unify that layer also? Let me know.
Thanks

**Comment by cloud-fan** [comment]
Yea let's unify that as well

**Comment by sujith71955** [comment]
> Yea let's unify that as well

sure

**Comment by sujith71955** [comment]
Retest this please.

**Review by cloud-fan — APPROVED** [review]

**Comment by sujith71955** [comment]
@cloud-fan @dongjoon-hyun  Thanks for your valuable time and guidance :)

**Comment by dongjoon-hyun** [comment]
Ur, @sujith71955 
The test failure is relevant to this PR. Could you check once more?
```
org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
```

**Comment by sujith71955** [comment]
> Ur, @sujith71955
> The test failure is relevant to this PR. Could you check once more?
> 
> ```
> org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
> ```

Will check and update the PR. thanks

**Comment by sujith71955** [comment]
> > Ur, @sujith71955
> > The test failure is relevant to this PR. Could you check once more?
> > ```
> > org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
> > ```
> 
> Will check and update the PR. thanks

Issue got resolved, we cannot remove the column name validation logic from RelationConversion rule as this rule will run before 'PreprocessTableCreation`, `PreprocessTableInsertion`, `DataSourceAnalysis` and `HiveAnalysis` rules. so before converting the relation we shall do the validation  as per the existing logic.

**Comment by sujith71955** [comment]
@cloud-fan @dongjoon-hyun  Testcases passed now, let me know for any clarifications/suggestions. thanks

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**